redsky/sinks.py

```diff
@@ -0,0 +1,99 @@
+"""Module for automatically storing processed data in a databroker"""
+
+
+##############################################################################
+#
+# redsky            by Billinge Group
+#                   Simon J. L. Billinge sb2896@columbia.edu
+#                   (c) 2017 trustees of Columbia University in the City of
+#                        New York.
+#                   All rights reserved
+#
+# File coded by:    Christopher J. Wright
+#
+# See AUTHORS.txt for a list of people who contributed.
+# See LICENSE.txt for license information.
+#
+##############################################################################
+
+
+class StoreSink(object):
+    """Sink of documents to save them to the databases.
+
+    The input stream of (name, document) pairs passes through unchanged.
+    As a side effect, documents are inserted into the databases and external
+    files may be written.
+
+    Parameters
+    ----------
+    db: ``databroker.Broker`` instance
+        The databroker to store the documents in, must have writeable
+        metadatastore and writeable filestore if ``external`` is not empty.
+    external_writers : dict
+        Maps data keys to a ``WriterClass``, which is responsible for writing
+        data to disk and creating a record in filestore. It will be
+        instantiated (possible multiple times) with the argument ``db.fs``.
+        If it requires additional arguments, use ``functools.partial`` to
+        produce a callable that requires only ``db.fs`` to instantiate the
+        ``WriterClass``.
+        """
+
+    def __init__(self, db, external_writers=None):
+        self.db = db
+        if external_writers is None:
+            self.external_writers = {}  # {'name': WriterClass}
+        else:
+            self.external_writers = external_writers
+        self.writers = None
+
+    def __call__(self, nd_pair):
+        name, doc = nd_pair
+        name, doc, fs_doc = getattr(self, name)(doc)
+        # The mutated fs_doc is inserted into metadatastore.
+        fs_doc.pop('filled', None)
+        fs_doc.pop('_name', None)
+        self.db.mds.insert(name, fs_doc)
```

Comment:
Put a try here, in case we try to put the same document in twice?
If the try fails, do not process any other documents associated with that start (I assume that it fails on the start, it would be odd to have different start uids but the same descriptor uid)